man/runlevel.8
.TH ninit-runlevel 8 "Jan 16, 2010"
.SH NAME
ninit-runlevel 
\- set runlevels for compatibility with SysVinit
.SH SYNOPSIS
.B ninit-runlevel
.B LEVEL
[OPTION]... [-] [NAME=VALUE]... [PROGRAM [ARG]...]
.br
.B ninit-runlevel
.
.SH DESCRIPTION
.B ninit-runlevel
is a helper program to start service (script, program) with the same
environ as SysVinit.  It also modifies the files
.I /var/run/utmp
and
.I /var/log/wtmp
writing runlevel records in them.  
It is used mainly in ninit-inittab(8) output script.

It sets the following variables:
INIT_VERSION=2.86,
CONSOLE=/dev/console,
RUNLEVEL,
PREVLEVEL,
PATH=/bin:/usr/bin:/sbin:/usr/sbin.
The variable NINIT_RUNLEVEL always has the same value
as RUNLEVEL.
The program does not modify or reset the variable INIT_HALT.

.SH OPTIONS
Actually  
.B ninit\-runlevel
has the same syntax between 
.B LEVEL
and
.B PROGRAM
as the command env(1).

If 
.B ninit-runlevel
is started as a service you can use 
.I environ 
file 
to set some environment variables.
However in this case 
.B ninit-runlevel
overwrites the variables:
INIT_VERSION,
CONSOLE,
RUNLEVEL,
PREVLEVEL,
PATH
and NINIT_RUNLEVEL.
You can reset some of them, for example PATH.
.TP
\-i
start with an empty environment
.br
a mere - implies -i
.TP
\-u 
.I variable
.br
remove the
.I variable 
from the environment

.SH "EXAMPLES"
In the next examples we modify PATH, INIT_HALT and CONSOLE

.B ninit-runlevel S 
PATH=/sbin:/bin /etc/rc.d/rc.single
.br
.B ninit-runlevel 0 
INIT_HALT=POWERDOWN -u CONSOLE /etc/rc.d/rc.halt

See the environ with (don't start it as root \-
the program will then modify 
.I /var/run/utmp
)

.B ninit-runlevel 5
/usr/bin/env

.SH WARNING
Don't remove the file 
. I /var/run/utmp
if you want correct PREVLEVEL and RUNLEVEL variables.
.B ninit-runlevel
stores the info about levels there.
.B ninit 
does not use the
.I  /var/run/utmp
file, nor PREVLEVEL, RUNLEVEL, NINIT_RUNLEVEL.

.SH "FILES"
.I /var/run/utmp
.br
.I /var/log/wtmp

.SH "SEE ALSO"
utmp(5), env(1), runlevel(8)
.br
ninit(8), ninit\-sysvinit(8), ninit\-shutdown(8), ninit\-inittab(8)
.br
init(1), shutdown(8), reboot(8), halt(8)


CHANGES
0.14:   2010-01-19 00:27:01
  added nkillall.8  
  gid is not default anymore.  see ninitfeatures.h
  uid supports 22:33:100:200:300
  nkillall: \e escape
  shutdown: set cron off
  nkillall: signals with letters like nsvc
  shutdown and nkillall: makes more sync(2), flag -q
  added THANKS ;-)
  added helpers: remove procfs;  thanks to them snvc is smaller
  nsvc options -onumber -unumber; option -E
  Cron has option -C
  /etc/ninit/.nsvc_help
  nsvc, shutdown, reload uses variables NINIT_MEMORY and NINIT_HOME
  nsvc: option -S
  shutdown: option -s
  manpages for all programs in /sbin ;-)
  update manpages (Boris Grozev)
  Makefile: man_txt
  serdo.c: read only one file again
  better sleep in open_inout.h
  serdo.c improvments; buildin: killall5, simple echo
  nsvc.c: -C+number
  scripts/conf
  ninit.8: NONROOT USAGE, using /sbin/init with ninit
  ninit-scan: (Wayne Marshall)

0.13.7:   2009-12-16 10:47:50
  bootlog.c (flush_root)
  0.13.1.4 renamed to 0.13.2
  service start 'end' (if it exixts; X_OK) after finishing
  fixed a bug in shutdown.c (thx to Jan)
  With flag -s ninit-shutdown don't start halt and reboot services
  shutdown.c:  improvements; flags -S -q
  shutdown.c:  removed flags -s -q; new flags -E -v -T
  added nkillall
  shutdown.c:  options -m and -E work together
  nkillall print escape like echo

0.13.1:   2009-06-27 12:59:38
  updated bootlog
  Makefile logs 'make tests' in tests_log using boolog
  fixing quotes in serdo

0.13:   2009-01-10 18:11:42
  it's possible to set parameters in wait.  example: some_service:180:3
  install-bin print help and string errors
  better sync mode.  maxsync is removed.  echo 200 > sync
  updated scripts/ninit_test.sh
  comments (#) in depends, params, environ, wait, softlimit
  install-bin (verbose mode)
  nsvc -V
  errmsg_put.c; buffer_*.c
  setup, rsetup, sys-rsetup starts with args: $1=service $2=service_pid
  EXTRACT_* trick (automatic create header files)
  stuct process is 20 bytes on x86_64 (see struct_root.h)
  changed BIFFER_INIT macro
  utmp_io.c
  Makefile: install_other
  serdo is intslled only with: make install_other
  t_write.h uses PIPE_BUF (thx Laurent Bercot)
  nsvc.8 improvements
  ninit_test.sh creates services in ./etc/ninit/
  ninit-huge
  added printf.c and ninit.spec
  make FLAG_DEBUG=no
  removed flag X_OK in some syscalls access.
  better addprocess

0.12.1:   2008-01-03 17:15:08
  ninit-depends.  convert directory to file (depends.dir -> depends)
  removed unused stat.h in headers
  updated serdo (static)
  two environ vars: NINIT_HOME, INIT_HOME.  (Thanks to Stamatis Mitrofanis)
  ninit logo: /etc/ninit/.sync
  install-bin uses chown32 instead chown on some systems
  err.c and err_b.c uses the macros va_start and va_arg
  fu.c (u &= 0xffffffff)

0.12:
  almost all arrays are moved in stack.  small data and bss sections.
  assembler functions (i386) for some DJB functions.
  modifications in Makefile.
  100 Euro security guarantee.

0.11.2:
  ninit-reload accept agument -e (change environ) -E file_env
  program ~/sys/update
  service flags: pause, pause-wait
  service flag cron:  a:b[:c]
    can contain many lines.
  flag -K for nsvc.
  pidfilehack --> ninit-pidfile.
  make i386 -- build static daemons (don't use dietlibc)
  Makefile is smaller and simpler
  gcc-4.1.2 works now.
  install-bin
  stuct mem in initialized in the main program.

0.11.1:
  ninit-reload accepts args (-r3 -r5 -r33 ...)
  ninit-inittab sets the variables NINIT_RUNLEVEL, INIT_VERSION
  if /etc/minit/{in|out} FIFO's exist then make symbolic links.
  new run flags: sys-rsetup, pidfile
  if the service sysvinit exist, ninit create /dev/initctl 
  ninit-reload and nsvc checks for environ NINIT_HOME
  nsvc -[drR] ALL - change all services.
    Change service ALL with: nsvc -[drR] ops ALL
  new program: ninit-sysvinit.  configuration file: sysvinit-timeout
    fimeout[:fork-mode].  example:  echo 600:0 > sysvinit-timeout
  ninit catch signal SIGPWR
  new run flags: alarm, maxsync
  nsvc output is different
  shell script ninit-service
  new program run-wait

0.11:
  Sun Jun 17 22:08:22 EEST 2007
  I decided to rename all without changes.
  The name ninit must be spelled nano-init.

0.9.11:
  rewritten by Nikola Vladov

0.9.1:
  fix embarassing typo in msvc (Gelu G. Lupas)

0.9:
  See minit CHANGES - http://www.fefe.de/minit


runlevel.c
#include <unistd.h>
#include <utmp.h>
#include <time.h>
#include <alloca.h>
#include <fcntl.h>
#include <sys/utsname.h>
#include "ninitfeatures.h"
#include "utmp_defs.h"

static unsigned long env_free, env_allocated;
#include "contrib/put_env.h"

static void init_utmp(struct utmp *u) {
  byte_zero(u, sizeof(struct utmp));
  str_copy(u->ut_line,"~");
  str_copy(u->ut_id,"~~");
  u->ut_tv.tv_sec=time(0);
}

int main(int argc, char **argv) {
  pid_t prev=0;
  int i,fd;
  char **env, ch, prevlevel[20], ninit_runlevel[20];
  char boot_time=0;
  struct utmp u;
  struct utsname uname_buf;
  off_t pos=0, runlevel_pos=0;

  str_copy(prevlevel, "PREVLEVEL=N");
  str_copy(ninit_runlevel,  "NINIT_RUNLEVEL=N");

  fd=open(_PATH_UTMP, (argc>1) ? O_RDWR : O_RDONLY);
  while (utmp_io(fd, &u, F_RDLCK)) {
    if (u.ut_type == BOOT_TIME) boot_time = 1;
    if (u.ut_type == RUN_LVL) {
      prev = u.ut_pid;
      runlevel_pos = pos;

      if (argc < 2) {
	char m[] = "N N\n";
	i = prev / 256; if (i) m[0] = i;
	i = prev % 256; if (i) m[2] = i;
	write(1, m, 4);
	return 0;
      }
    }
    pos += sizeof(struct utmp);
  }

  if (argc < 2) _exit(1);
  prev %= 256;
  ch = argv[1][0];

  if (boot_time == 0) {
    init_utmp(&u);
    u.ut_type=BOOT_TIME;
    str_copy(u.ut_user,"reboot");
    if (lseek(fd,pos,SEEK_SET) == pos)
      utmp_io(fd,&u,F_WRLCK);
  }

  pos = runlevel_pos;
  if (pos == 0 && lseek(fd,0,SEEK_SET) == 0) {
    while (utmp_io(fd, &u, F_RDLCK)) {
      if (u.ut_type == RUN_LVL) break;
      pos += sizeof(struct utmp);
    }
  }

  init_utmp(&u);
  u.ut_type=RUN_LVL;
  str_copy(u.ut_user,"runlevel");
  if (prev==0) prev='N';
  u.ut_pid = (256*prev) + ch;

  if (lseek(fd,pos,SEEK_SET) == pos)
    utmp_io(fd,&u,F_WRLCK);
  close(fd);
    
  if (uname(&uname_buf) == 0)
    str_copyn(u.ut_host, uname_buf.release, 32); /* XXX overflow ? */
  do_wtmp(&u);
  
  /* ------ exnivon + exeve ----- */
  if (environ==0) environ = argv+argc;
  for (env=environ; *env; ++env) env_allocated++;

  env_free = argc + 12;
  env=alloca((env_allocated + env_free + 1) * sizeof(char*));
  if (env==0) return -1;
  byte_copy(env, (env_allocated+1)*sizeof(char *), environ);
  environ = env;
  
  prevlevel[10] = prev;
  ninit_runlevel[15] = ch;

  put_env(prevlevel);
  put_env(ninit_runlevel);
  put_env(ninit_runlevel+6);
  put_env("INIT_VERSION=" SYSVINIT_VERSION);
  put_env("CONSOLE=/dev/console");
  put_env("PATH=/bin:/usr/bin:/sbin:/usr/sbin");
  
  for (i=2; i<argc; ++i) {
    char *v=argv[i];
    char *nenv=0;
    if (v[0]=='-') {
      if (v[1]==0) {
	environ[0]=0;
      } else {
	switch (*++v) {
	case 'i': environ[0]=0; goto do_it;
	case 'u':
	  if (v[1]) { nenv = v+1; goto do_it; }
	  else if (argv[++i]) { nenv = argv[i]; goto do_it; }
	default:
	  goto ready;
	}
      }
    } else {
      if (v[str_chr(v,'=')] == 0) goto ready;
      nenv = v;
    }

  do_it:
    if (nenv) put_env(nenv);
    continue;
  }

 ready:
  argv += i;
  if (argv[0]) {
    execve(argv[0], argv, environ);
    return 100;
  }
  return 0;
}
